### ISSUE TITLE
Fda background process not ended when HEC-FDA software is closed.

### ISSUE THREAD

**Issue by jbslaughter** [HydrologicEngineeringCenter/HEC-FDA#370]
**Build:** CI-CD #41 (Merge pull request #354), dated 2022-05-20.

Closing the HEC-FDA software does not end all processes. This seems to happen after data is saved to the study DBF (e.g., new damage category created). Issue found after closing the HEC-FDA software and trying to rename the folder containing the created and edited study files. Had to use Task Manager to End the "Fda" background process.

**Comment by codymccoy** [comment]
@HenryGeorgist Deziree has encountered a similar experience. She had to open task manager and kill fda. This is a call stack that she gave me. Looks like something with TextFileMessageSubscriber?

![image](https://user-images.githubusercontent.com/47757848/171434894-1da60520-e016-4a90-9eee-b4c4da4efd93.png)

**Comment by rnugent3** [comment]
@HenryGeorgist any update on your thoughts on this issue?

**Comment by HenryGeorgist** [comment]
i suggested in chat that we need to close the messagehub and the texfile subscriber on the closing of the hec-fda view

**Comment by HenryGeorgist** [comment]
https://github.com/HydrologicEngineeringCenter/Framework/blob/9bee2706c670395a7268f080e7d50a6a73f598b7/Base/Implementations/MessageHub.cs#L165

unsubscribe all would allow all handlers to be removed.

**Comment by HenryGeorgist** [comment]
https://github.com/HydrologicEngineeringCenter/Framework/blob/9bee2706c670395a7268f080e7d50a6a73f598b7/View/Implementations/TextFileMessageSubscriber.cs#L128

dispose allows the textfile subscriber to be disposed

all this needs to happen in the closing of the main window - which i think is handled in this class?

https://github.com/HydrologicEngineeringCenter/HEC-FDA/blob/main/View/ViewWindow.xaml.cs

you could also do a closing event in app.xaml

**Comment by codymccoy** [comment]
I will attempt the approach described by Will. I am not seeing this issue on my computer. I can save data in fda, close fda, then rename the study folder. This will make it tough for me to test if the new code has solved the problem. I think we will want Julia and Deziree to see if they see this issue again.

**Comment by HenryGeorgist** [comment]
the issue seems to be on file options associated with creating the textfile subscriber log file... maybe it is unrelated all together.

**Comment by codymccoy** [comment]
I added these first two lines to the window_closing event and it throws an exception when closing.
![image](https://user-images.githubusercontent.com/47757848/173155723-4aff8202-f7d2-4dc8-ab26-040e43fe3493.png)

**Comment by HenryGeorgist** [comment]
in message hub it needs to be a for i = loop not a foreachloop

you cannot remove in a foreach apparently

**Comment by HenryGeorgist** [comment]
or actually, it is probably possible to just remove the loop and do a remove - lists should be dynamic and be able to remove interenally.

**Comment by rnugent3** [comment]
Hi all. There was a question last time we met whether this was fixed. I can confirm that this is not fixed. Working on a build off main, I launched the executable, opened a study, then closed the software. I re-launched, attempted to open the same study, and could not because a process was active on the log file. I think this needs to be a priority for us.

**Comment by RyanM-RMA** [comment]
I resolved this in the following changes: 
69e6fa7f9cfba1f86fafd99f9384ed6756e31f80
e1edc2e7767f1b2f9c816d1d231c74c9f53ac1dc

The code I submitted is effectively a patch for the larger issue.  The issue is that the main window is not shutting down the main thread as it's supposed to.  I don't see anything in particular preventing it from shutting down the main thread after destruction, and there aren't any foreground threads running, only dead background workers that are most likely alive because they're part of a pool.

As I told Cody, I believe we need to consider separating the main window and the subsequent windows that pop up.  I believe this could be part of the issue.  It could also potentially be a memory leak with the window itself somewhere, but I have my doubts about that.

**Comment by rnugent3** [comment]
thank you @RyanM-RMA !!!!! do you have a ball park guess on what it would take to separate the windows?

**Comment by RyanM-RMA** [comment]
@codymccoy is more familiar with the design involving the ViewWindow class, so he would be better at estimating that.

**Comment by rnugent3** [comment]
Another contract was my thought, too. Let's keep this in backlog.

**Comment by HenryGeorgist** [comment]
i think the best solution is to use app_startup in the app.xaml.  hijack its closing event to dispose of handlers. it isnt the behavior of the window to manage the singleton logger, it is the behavior of the whole application. it should get initiated in app_startup and disposed of in app_closing.

**Comment by rnugent3** [comment]
I am experiencing this issue again. I launch from the build folder. Close and attempt to rebuild. I cannot rebuild because the stuff happening in the background has not terminated.

**Comment by rnugent3** [comment]
I just ran into an error trying to clean the solution. I think that the error mentioned in the message below is probably related to this issue. 
`5>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(5599,5): warning MSB3061: Unable to delete file "C:\Code\C#\FDA\HEC-FDA\Build\FDA2.0\Importer.dll". Access to the path 'C:\Code\C#\FDA\HEC-FDA\Build\FDA2.0\Importer.dll' is denied.`

**Comment by rnugent3** [comment]
yup.... 
![image](https://user-images.githubusercontent.com/25163165/189457575-90d9d80b-c9fe-4ff2-b950-9e7988a70f2b.png)

@Brennan1994 @HenryGeorgist do you think this problem should be solved before we ship the beta - or just give the beta testers a heads up that they'll have to quit fda using the task manager?

**Comment by rnugent3** [comment]
@codymccoy please take the lead on this issue.

**Comment by codymccoy** [comment]
@Brennan1994 Are you able to reproduce this? I can't reproduce this. I tried opening multiple studies in different orders and never had anything locked. I feel like there might be different things happening throughout this github issue. It seems to have been the TextFileMessageSubscriber that was locking fda but the latest image posted by Richard is indicating that the issue is with the Importer.dll. So maybe the locking has to do with importing data from old fda? I tested creating studies by importing from old fda and didn't encounter the locking.

**Comment by rnugent3** [comment]
I have attempted to reproduce and cannot. I have tried a couple of different things that might have caused the issue. I think for me, I might have run into a problem that looked like this issue but is probably wholly unrelated to this issue. I have a study directory that I have attempted to delete and am told that the folder no longer exists every time I attempt to delete. So maybe when I attempt to open the study in that directory, a similar message gets sent to the software prompting your validation. 

@jbslaughter have you run into this issue at all lately? If not, we'll close the issue and I'll do a spiritual dance around the HEC patio to shun bad background process juju.
